### ISSUE TITLE
Distributed domain access from record destructor seg faults

### ISSUE THREAD

**Issue by ben-albrecht** [chapel-lang/chapel#5627]
### Summary of Problem

Accessing a distributed domain (or any array declared over that domain) from within a record destructor results in a segfault for `CHPL_COMM=none` and a `dereference to nil` error for `CHPL_COMM=gasnet`.

This error did not occur in version 1.14.0.

### Steps to Reproduce

**Source Code:**

```chapel
// Error occurs for any distribution, just picking Block as a simple example
use BlockDist;

const blockDomain = {1..1} dmapped Block(boundingBox={1..1});

const Domain = {1..1};

record Test {
  proc ~Test() {
    // nil by the time we call this deconstructor
    writeln(blockDomain);
    // still exists
    writeln(Domain);
  }
}

var t : Test;
```

**Compilation & Execution:**

 On master today (from `a9a5e95814a6` / #4762 and on):
```bash
 > chpl --comm=gasnet foo.chpl && ./a.out -nl 1

  $CHPL_HOME/modules/internal/ChapelArray.chpl:2350: error: attempt to dereference nil

 > chpl --comm=none foo.chpl && ./a.out

  Segmentation Fault: 11
```

On Chapel 1.14.0 (up to `d25e9279a127`)

```bash
 > chpl --comm=gasnet foo.chpl && ./a.out -nl 1

  0
  0

 > chpl --comm=none foo.chpl && ./a.out

  0
  0
```

**Associated Future Test(s):**

*TODO*

### Configuration Information

- Output of `chpl --version`: `chpl Version 1.14.0.ee1d3fd` (HEAD of master)

**Comment by bradcray** [comment]
Huh, weird...  Have you tried git bisecting to find the offending merge?

(PS -- Real developers use 'noinit'...  :)  Not that I expect that to fix this issue)

**Comment by ben-albrecht** [comment]
Thanks @npadmana for helping me track down this bug from a higher-level MPI module bug (#5573).

@mppf - it looks like this bug began after #4762 was merged.

**Comment by cassella** [comment]
``t`` and ``blockDomain`` (and ``Domain``) all go out of scope "at the
same time".

I can reproduce the failure, but it works for me if I change the last
line to

    { var t : Test; }

to ensure that ``t`` goes out of scope first.

**Comment by ben-albrecht** [comment]
@cassella -- this fixes the single locale version, but the gasnet version continues to dereference nil

**Comment by cassella** [comment]
Hmm.  That's not what I see.  (Now that I go back and build a
CHPL_COMM=none compiler, I don't see any failures there, with or
without the extra scope).

    fortytwo@magrathea:~/src/chpl$ ./dscope-noscope-local 
    {1..1}
    {1..1}

    fortytwo@magrathea:~/src/chpl$ ./dscope-extrascope-local 
    {1..1}
    {1..1}

    fortytwo@magrathea:~/src/chpl$ ./dscope-noscope-gasnet -nl 1
    $CHPL_HOME/modules/internal/ChapelArray.chpl:1582: error: attempt to dereference nil

    fortytwo@magrathea:~/src/chpl$ ./dscope-extrascope-gasnet -nl 1
    {1..1}
    {1..1}

Each of the -gasnet versions produced the same results for -nl 2
through -nl 4 as it did for -nl 1.


ChapelArray.chpl:1582 (my error) and :2350 (your error) are both
writeThis() procs, in "record _domain" and "record _array"
respectively.


I'm using the latest master, chpl Version 1.14.0.cc1b041

**Comment by ben-albrecht** [comment]
My mistake. I can confirm that additional scoping does work. 

However, the original example should still be valid. While `t` and `blockDomain`, and `Domain` all go out of scope "at the same time", their destructors should be called in the reverse order of when they were initialized.

**Comment by bradcray** [comment]
I forgot to confirm in yesterday's meeting whether @mppf considered himself to own this or not.  That said, prior to editing the previous comment, @ben-albrecht alleged that destructors were being called in the wrong order which could open this up to a number of other people to look into it.

**Comment by mppf** [comment]
I'm open to looking at it (after feature freeze) but I don't think it's obviously mine.

**Comment by bradcray** [comment]
Meaning you think it's a pre-existing condition that your PR exposed rather than something that it introduced?

**Comment by mppf** [comment]
I don't know. Clearly, the problem started with my PR, but it could be a bug that was revealed by it. After all, we weren't freeing distributed arrays/domains before that PR, so even if the order was wrong it probably would have worked...

**Comment by vasslitvinov** [comment]
I will look at this for code freeze, unless Michael beats me to it.

**Comment by vasslitvinov** [comment]
Curiously, I am the one who edited the relevant code in insertGlobalAutoDestroyCalls()
most recently. (Yes, in 2012.)

**Comment by vasslitvinov** [comment]
Since 2009, we insertGlobalAutoDestroyCalls() essentially in the order that the globals are parsed.

For example, for this program:

```chapel
use UsedPackage;
var UserVar1: UserRecord;
var UserVar2: UserRecord;
```

here are the module files and global variables therein, as of today:

```
chpl --print-module-files UserProgram.chpl

Parsing module files:
  modules/internal/ChapelBase.chpl
  modules/internal/ChapelStandard.chpl
  modules/internal/PrintModuleInitOrder.chpl
  modules/internal/startInitCommDiags.chpl
  modules/internal/CPtr.chpl
  modules/internal/CString.chpl
  modules/internal/String.chpl
  modules/internal/ChapelDebugPrint.chpl
  modules/internal/ChapelEnv.chpl
  modules/internal/MemConsistency.chpl
  modules/internal/Atomics.chpl
  modules/internal/NetworkAtomics.chpl
  modules/internal/NetworkAtomicTypes.chpl
  modules/internal/AtomicsCommon.chpl
  modules/internal/ChapelIteratorSupport.chpl
  modules/internal/ChapelThreads.chpl
  modules/internal/ChapelThreadsInternal.chpl
  modules/internal/ChapelTasksInternal.chpl
  modules/internal/ChapelTuple.chpl
  modules/internal/ChapelRange.chpl
  modules/internal/ChapelReduce.chpl
  modules/internal/ChapelSyncvar.chpl
  modules/internal/localeModels/flat/LocaleModel.chpl
chpl_emptyLocaleSpace   // domain
chpl_emptyLocales       // array over the above (!!)
  modules/internal/ChapelLocale.chpl
  modules/internal/DefaultRectangular.chpl
defaultDist            // no-op deinit
  modules/internal/LocalesArray.chpl
Locales                // alias
LocaleSpace            // domain
  modules/internal/ChapelArray.chpl
numPrivateObjects
_OpaqueIndexNext
  modules/internal/ChapelDistribution.chpl
chpl_emptyLocaleSpace   // domain  yes, different than LocaleModel's one
chpl_emptyLocales       // array over the above
  modules/internal/ChapelIO.chpl
  modules/internal/LocaleTree.chpl
chpl_localeTree   // no-op deinit
  modules/internal/DefaultAssociative.chpl
  modules/internal/DefaultSparse.chpl
  modules/internal/DefaultOpaque.chpl
  modules/internal/ChapelTaskID.chpl
  modules/internal/tasktable/off/ChapelTaskTable.chpl
  modules/internal/MemTracking.chpl
memLog  // string
  modules/internal/ChapelUtil.chpl
  modules/internal/ChapelError.chpl
  modules/internal/ChapelDynDispHack.chpl
  modules/internal/stopInitCommDiags.chpl
  modules/internal/StringCasts.chpl
  modules/internal/LocaleModelHelpFlat.chpl
  modules/internal/LocaleModelHelpMem.chpl
  modules/internal/ArrayViewSlice.chpl
  modules/internal/ArrayViewRankChange.chpl
  modules/internal/ArrayViewReindex.chpl
  modules/internal/LocaleModelHelpSetup.chpl
  modules/internal/LocaleModelHelpRuntime.chpl
  modules/internal/ChapelNumLocales.chpl
  UserProgram.chpl
UserVar1  // could have deinits
UserVar2
  modules/packages/UsedPackage.chpl
VarInUsedPackageModule  // could have a deinit
  modules/standard/Assert.chpl
  modules/standard/Types.chpl
  modules/standard/Math.chpl
  modules/standard/CommDiagnostics.chpl
  modules/standard/gen/linux64-gnu/SysCTypes.chpl
  modules/dists/DSIUtil.chpl
  modules/packages/Sort.chpl
defaultComparator   // no-op deinits
reverseComparator
  modules/standard/Reflection.chpl
  modules/standard/List.chpl
  modules/standard/SysBasic.chpl
  modules/standard/IO.chpl
_r   // these are strings
_rw
_cw
_cwr
_arg_to_proto_names  // tuple of strings
stdin    // channel's deinit is no-op, I think
stdout
stderr
  modules/packages/RangeChunk.chpl
  modules/packages/Search.chpl
  modules/standard/Sys.chpl
  modules/standard/SysError.chpl
  modules/standard/Regexp.chpl
```

**Comment by vasslitvinov** [comment]
So, what order **should** we destruct the globals?

* Globals in a given module destruct in reverse declaration order.
* Globals in user code destruct before libraries.
* If user main module `uses` other user module(s), in which order should those modules be processed?

For internal and standard modules, I think that these should be destructed last, in this order:

* locale-related globals, esp. arrays then domains
* I/O stuff
* all strings

Given the current situation (see my previous comment), this "internal/standard module" list already happens in roughly this order, so we could actually leave that order as-is. At least, until we have a reason to change it. If so, we only need to take care of user modules and perhaps package modules.

**Comment by bradcray** [comment]
My gut reaction is that modules should deinit themselves in the reverse order that they were initialized (which is determined by a DFS search on use clauses, IIRC).  Similarly, I think module-scoped variables should be deinit'd in the reverse order that they were initialized.

While looking at this, keep in mind that there's a longstanding feature request for users to be able to define their own deinit() routines for modules so that they can clean up things themselves as necessary (e.g., free global classes).  I've always hypothesized that that should be really trivial to drop in, particularly if we were cleaning up modules properly, but have never found the time to look into it.

**Comment by bradcray** [comment]
I'm not certain, but I think that compiler/passes/addInitGuards.cpp is where at least some of this module initialization order stuff is established -- noting it in case it's helpful.
